### PR TITLE
feat: search text from selection

### PR DESCRIPTION
**Pull request by mahaker**
close https://github.com/kbysiec/vscode-search-everywhere/issues/31
Provide a more similar feel to Intellij's IDE.

Sorry for the delay in send the PR.

### CODE REVIEW

File: src/test/testSetup/controller.testSetup.ts [hunk 1]
```diff
@@ -486,6 +490,62 @@ export const getTestSetups = () => {
...
+          },
+          {
+            object: quickPick,
+            method: "setText",
+          },
+          {
+            object: config,
+            method: "fetchShouldQuickPickOpenFromSelection",
+            returns: true,
+          },
```
Comment (mahaker) [inline src/test/testSetup/controller.testSetup.ts:516]: Could not stub `vscode.window.activeTextEditor`.
So I could not write a test when `vscode.window.activeTextEditor` is undefined 😭 

At that time I had to write code like this.
```js
{
  object: vscode,
  method: "window",
  returns: ({ activeTextEditor: undefined })
},
```

File: src/controller.ts [hunk 2]
```diff
@@ -111,6 +121,13 @@ function shouldLoadDataFromCacheOnStartup() {
   );
 }
 
+function shouldQuickPickOpenFromSelection() {
+  return (
+    quickPick.isInitialized() &&
+    fetchShouldQuickPickOpenFromSelection()
+  );
+}
+
```
Comment (kbysiec) [inline src/controller.ts:130]:
```suggestion
function shouldSearchSelection(editor: vscode.TextEditor | undefined) {
  return quickPick.isInitialized() && fetchShouldSearchSelection() && !!editor;
}
```

File: src/controller.ts [hunk 3]
```diff
@@ -21,6 +22,15 @@ import {
...
   quickPick.show();
+
+  const editor = vscode.window.activeTextEditor;
+  if (
+    shouldQuickPickOpenFromSelection() &&
+    editor !== undefined
+  ) {
+    const { start, end } = editor.selection;
+    quickPick.setText(editor.document.getText(new vscode.Range(start, end)));
+  }
```
Comment (kbysiec) [inline src/controller.ts:33]:
```suggestion

  const activeEditorOrUndefined = vscode.window.activeTextEditor;
  
  if (controller.shouldSearchSelection(activeEditorOrUndefined)) {
    const activeEditor = activeEditorOrUndefined as vscode.TextEditor;
    
    const { start, end } = activeEditor.selection;
    quickPick.setText(editor.document.getText(new vscode.Range(start, end)));
  }
```

ShouldSearchSelection method be invoked from controller due to necessity to mock it in test correctly.

File: src/controller.ts [hunk 4]
```diff
@@ -219,6 +236,7 @@ export const controller = {
   shouldLoadDataFromCacheOnQuickPickOpen,
   shouldIndexOnStartup,
   shouldLoadDataFromCacheOnStartup,
+  shouldQuickPickOpenFromSelection,
```
Comment (kbysiec) [inline src/controller.ts:239]:
```suggestion
  shouldSearchSelection,
```

File: src/test/util/mockFactory.ts [hunk 5]
```diff
@@ -64,6 +64,7 @@ export const getConfiguration = (): { [key: string]: any } => {
       exclude: ["**/node_modules/**"],
       excludeModee: ExcludeMode.SearchEverywhere,
       shouldWorkspaceDataBeCached: true,
+      shouldQuickPickOpenFromSelection: true,
```
Comment (kbysiec) [inline src/test/util/mockFactory.ts:67]:
```suggestion
      shouldSearchSelection: true,
```

File: src/test/test/config.test.ts [hunk 6]
```diff
@@ -171,4 +171,16 @@ describe("Config", () => {
       );
     });
   });
+
+  describe("fetchShouldQuickPickOpenFromSelection", () => {
```
Comment (kbysiec) [inline src/test/test/config.test.ts:175]:
```suggestion
  describe("fetchShouldSearchSelection", () => {
```

File: src/test/test/config.test.ts [hunk 7]
```diff
@@ -171,4 +171,16 @@ describe("Config", () => {
...
+    it("1: should return boolean from configuration", () => {
+      const section = "searchEverywhere";
+      const key = "shouldQuickPickOpenFromSelection";
```
Comment (kbysiec) [inline src/test/test/config.test.ts:178]:
```suggestion
      const key = "fetchShouldSearchSelection";
```

File: src/test/test/config.test.ts [hunk 8]
```diff
@@ -171,4 +171,16 @@ describe("Config", () => {
...
+
+      assert.equal(
+        config.fetchShouldQuickPickOpenFromSelection(),
```
Comment (kbysiec) [inline src/test/test/config.test.ts:181]:
```suggestion
        config.fetchShouldSearchSelection(),
```

File: src/test/test/controller.test.ts [hunk 9]
```diff
@@ -139,6 +139,26 @@ describe("controller", () => {
...
+
+    it(`11: should quickPick.show and quickPick.loadItems invoked, and quickOpen.setText method not be invoked
+      if shouldQuickPickOpenFromSelection method return false`, async () => {
+      const [loadItemsStub, showStub, setTextStub] = setups.search11();
+      await controller.search();
+
+      assert.equal(loadItemsStub.calledOnce, true);
+      assert.equal(showStub.calledOnce, true);
+      assert.equal(setTextStub.calledOnce, false);
+    });
```
Comment (kbysiec) [inline src/test/test/controller.test.ts:161]:
```suggestion

    it(`10: should set selected text as quickPick text if shouldSearchSelection
      returns true`, async () => {
      const [setTextStub] = setups.search10();
      await controller.search();

      assert.equal(setTextStub.calledWith("test text"), true);
    });
    
    it(`11: should not invoke quickPick.setText method if shouldSearchSelection
      returns false`, async () => {
      const [setTextStub] = setups.search11();
      await controller.search();

      assert.equal(setTextStub.calledOnce, false);
    });
```

I would suggest to test only whether based on shouldSearchSelection method and shouldSearchSelection method test separately.

File: src/test/test/controller.test.ts [hunk 10]
```diff
@@ -510,6 +530,27 @@ describe("controller", () => {
     });
   });
 
+  describe("shouldQuickPickOpenFromSelection", () => {
```
Comment (kbysiec) [inline src/test/test/controller.test.ts:533]:
```suggestion
  describe("shouldSearchSelection", () => {
```

File: src/test/test/controller.test.ts [hunk 11]
```diff
@@ -510,6 +530,27 @@ describe("controller", () => {
...
+    it("1: should return false if quickPick is not initialized", () => {
+      setups.shouldQuickPickOpenFromSelection1();
```
Comment (kbysiec) [inline src/test/test/controller.test.ts:535]:
```suggestion
      setups.shouldSearchSelection1();
```

File: src/test/test/controller.test.ts [hunk 12]
```diff
@@ -510,6 +530,27 @@ describe("controller", () => {
...
+
+      assert.equal(controller.shouldQuickPickOpenFromSelection(), false);
```
Comment (kbysiec) [inline src/test/test/controller.test.ts:537]:
```suggestion
      assert.equal(controller.shouldSearchSelection(), false);
```

File: src/test/test/controller.test.ts [hunk 13]
```diff
@@ -510,6 +530,27 @@ describe("controller", () => {
...
+    });
+
+    it("2: should return false if fetchShouldQuickPickOpenFromSelection method returns false", () => {
```
Comment (kbysiec) [inline src/test/test/controller.test.ts:540]:
```suggestion
    it("2: should return false if fetchShouldSearchSelection method returns false", () => {
```

File: src/test/test/controller.test.ts [hunk 14]
```diff
@@ -510,6 +530,27 @@ describe("controller", () => {
...
+      setups.shouldQuickPickOpenFromSelection2();
```
Comment (kbysiec) [inline src/test/test/controller.test.ts:541]:
```suggestion
      setups.shouldSearchSelection2();
```

File: src/test/test/controller.test.ts [hunk 15]
```diff
@@ -510,6 +530,27 @@ describe("controller", () => {
...
+
+      assert.equal(controller.shouldQuickPickOpenFromSelection(), false);
```
Comment (kbysiec) [inline src/test/test/controller.test.ts:543]:
```suggestion
      assert.equal(controller.shouldSearchSelection(), false);
```

File: src/test/test/controller.test.ts [hunk 16]
```diff
@@ -510,6 +530,27 @@ describe("controller", () => {
...
+    });
+
+    it(`3: should return true if quickPick is already initialized,
+      fetchShouldQuickPickOpenFromSelection methods return true`, () => {
+      setups.shouldQuickPickOpenFromSelection3();
+
+      assert.equal(controller.shouldQuickPickOpenFromSelection(), true);
+    });
+  });
+
```
Comment (kbysiec) [inline src/test/test/controller.test.ts:553]:
```suggestion
    it("3: should return false if editor doesn't exist", () => {
      const editor = setups.shouldSearchSelection3();
      assert.equal(controller.shouldSearchSelection(editor), false);
    });

    it(`4: should return true if quickPick is initialized,
      fetchShouldSearchSelection returns false and editor exists`, () => {
      const editor = setups.shouldSearchSelection4();
      assert.equal(controller.shouldSearchSelection(editor), true);
    });
  });

```

Add one more test to verify if active editor exists.

File: src/test/testSetup/controller.testSetup.ts [hunk 17]
```diff
@@ -1032,6 +1092,56 @@ export const getTestSetups = () => {
...
+        ],
+        sandbox
+      );
+    },
```
Comment (kbysiec) [inline src/test/testSetup/controller.testSetup.ts:1144]:
```suggestion
    //above getTestSetups
    function stubComponentsForShouldSearchSelection(
      sandbox: sinon.SinonSandbox,
      returnsValues: { [key: string]: any }
    ) {
      return stubMultiple(
        [
          {
            object: quickPick,
            method: "isInitialized",
            returns: returnsValues.isInitialized,
          },
          {
            object: config,
            method: "fetchShouldSearchSelection",
            returns: returnsValues.fetchShouldSearchSelection,
          },
        ],
        sandbox
      );
    }

    shouldSearchSelection1: () => {
      stubComponents(sandbox);
      stubComponentsForShouldSearchSelection(sandbox, {
        isInitialized: false,
        fetchShouldSearchSelection: true,
      });
      const editor = getTextEditorStub(true, "test text");

      return editor;
    },
    shouldSearchSelection2: () => {
      stubComponents(sandbox);
      stubComponentsForShouldSearchSelection(sandbox, {
        isInitialized: true,
        fetchShouldSearchSelection: false,
      });
      const editor = getTextEditorStub(true, "test text");

      return editor;
    },
    shouldSearchSelection3: () => {
      stubComponents(sandbox);
      stubComponentsForShouldSearchSelection(sandbox, {
        isInitialized: true,
        fetchShouldSearchSelection: true,
      });
      const editor = undefined;

      return editor;
    },
    shouldSearchSelection4: () => {
      stubComponents(sandbox);
      stubComponentsForShouldSearchSelection(sandbox, {
        isInitialized: true,
        fetchShouldSearchSelection: true,
      });
      const editor = getTextEditorStub(true, "test text");

      return editor;
    },
```

You can find above mock for undefined return for editor.

File: src/test/testSetup/controller.testSetup.ts (hunk 1 above)
Comment (kbysiec) [inline src/test/testSetup/controller.testSetup.ts:516]:
```suggestion
```

Look at code suggestion in next code block.

File: src/test/testSetup/controller.testSetup.ts [hunk 18]
```diff
@@ -486,6 +490,62 @@ export const getTestSetups = () => {
...
+          },
+          {
+            object: config,
+            method: "fetchShouldQuickPickOpenFromSelection",
+            returns: false,
+          },
+        ],
+        sandbox
+      );
+    },
```
Comment (kbysiec) [inline src/test/testSetup/controller.testSetup.ts:548]:
```suggestion
      search10: () => {
      stubComponents(sandbox);
      stubComponentsForSearch(sandbox, {
        isInitialized: true,
        shouldIndexOnQuickPickOpen: false,
        shouldLoadDataFromCacheOnQuickPickOpen: false,
        shouldSearchSelection: true,
      });
      const editor = getTextEditorStub(true, "test text");

      return stubMultiple(
        [
          {
            object: quickPick,
            method: "setText",
          },
          {
            object: quickPick,
            method: "loadItems",
          },
          {
            object: quickPick,
            method: "show",
          },
          {
            object: vscode.window,
            method: "activeTextEditor",
            isNotMethod: true,
            returns: editor,
          },
        ],
        sandbox
      );
    },
    search11: () => {
      stubComponents(sandbox);
      stubComponentsForSearch(sandbox, {
        isInitialized: true,
        shouldIndexOnQuickPickOpen: false,
        shouldLoadDataFromCacheOnQuickPickOpen: false,
        shouldSearchSelection: false,
      });

      return stubMultiple(
        [
          {
            object: quickPick,
            method: "setText",
          },
          {
            object: quickPick,
            method: "loadItems",
          },
          {
            object: quickPick,
            method: "show",
          },
        ],
        sandbox
      );
    },
```

**Review by kbysiec — CHANGES_REQUESTED** [review]
Hi @mahaker,

First of all thank you for your PR and apologies for such a delay in reviewing and responding. I had difficult period with no time for side projects at all.

I have reviewied your solution and changed in some places because I would like to be consistent with the whole codebase. 
Additionally, I can see the one think to improve - if you open the quickPick with text selected and debounce enabled, there is a blinking of itemsList on open. But don't worry I have an idea how to fix it so just after merging your PR I will add some improvements from my side and release the new version.

Thanks one more time for your help.
In case of any questions, let me know.

Kamil

**Comment by mahaker** [comment]
Hi @kbysiec !

Thanks for the very thoughtful review!
I have committed the fixes and would appreciate your confirmation.

Best regards.

**Comment by kbysiec** [comment]
@mahaker,

thank you for your changes. I have reviewed them, merged and after some improvements on my side released in new version  [v.2.1.0](https://github.com/kbysiec/vscode-search-everywhere/tree/v.2.1.0)

**Pull request merged**